### PR TITLE
Dynamic highlight

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -383,6 +384,25 @@ function run () {
...
 
+  // ---------------- ContextualListener -----------------------
+  this._components.contextualListener = new ContextualListener({
+    getCursorPosition: () => {
+      return this._components.editor.getCursorPosition()
+    },
+    getCompilationResult: () => {
+      return compiler.lastCompilationResult
+    },
+    warnExpression: (position) => {
```
Comment (chriseth) [inline src/app.js]: Why is it called "warn"?

File: src/app/editor/contextualListener.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      this._warnExpressions(api.getCursorPosition(), api.getCompilationResult())
+    }, 1000)
+  }
+
+  _warnExpressions (cursorPosition, compilationResult) {
+    if (this.currentPosition === cursorPosition) return
+    this._stopWarning()
+    this.currentPosition = cursorPosition
+    if (compilationResult && compilationResult.data && compilationResult.source) {
+      var nodes = this.sourceMappingDecoder.nodesAtPosition(null, cursorPosition, compilationResult.data.sources[compilationResult.source.target])
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: I'm getting `this.sourceMappingDecoder.nodesAtPosition is not a function`

File: src/app/editor/contextualListener.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      if (nodes && nodes.length && nodes[nodes.length - 1]) {
+        this._warnExpression(nodes[nodes.length - 1], compilationResult)
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: `nodes.length - 1` is the one with the "smallest" source location, i.e. the deepest entry in the AST?

File: src/app/editor/contextualListener.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    if (compilationResult && compilationResult.sources) {
+      var self = this
+      var callback = {}
+      callback['*'] = function (node) {
+        if (node && node.attributes && node.attributes.referencedDeclaration) {
+          if (!self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration]) {
+            self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration] = []
+          }
+          self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration].push(node)
+          self._index['Expressions'].push(node)
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Why is it called "expressions"?

File: src/app/editor/contextualListener.js (hunk 3 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: yes

File: src/app.js (hunk 1 above)
Comment (yann300) [reply src/app.js]: Means that the module wants to warn app.js that this position need a particular treatment. cause we might extend this function to do something else than just highlight.

File: src/app.js [hunk 5]
```diff
@@ -383,6 +384,25 @@ function run () {
...
+      position = offsetToLineColumnConverter.offsetToLineColumn(position, position.file, compiler.lastCompilationResult)
+      return editor.addMarker(position, config.get('currentFile'), 'highlightcall')
```
Comment (chriseth) [inline src/app.js]: Why highlightcall? Should be something like "highlightCurrentDeclaration" or "...CurrentRefereces"

File: src/app/editor/contextualListener.js (hunk 4 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: in this context, `Expressions` would be a general term to describe use of a declared item.

File: src/app.js (hunk 5 above)
Comment (yann300) [reply src/app.js]: yes right

Comment (chriseth) [inline src/app.js]: Be careful - references across files are easily possible.

File: src/app/editor/contextualListener.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        ReferencedDeclarations: {},
+        Expressions: []
+      }
+      if (success) {
+        this._buildIndex(data, source)
+      }
+    })
+    this.sourceMappingDecoder = new SourceMappingDecoder()
+    this.astWalker = new AstWalker()
+    setInterval(() => {
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Is this cancelled if we recompile in the meantime?

Comment (chriseth) [reply src/app/editor/contextualListener.js]: (or change the text)?

File: src/app/editor/contextualListener.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      this._warnExpressions(api.getCursorPosition(), api.getCompilationResult())
+    }, 1000)
+  }
+
+  _warnExpressions (cursorPosition, compilationResult) {
+    if (this.currentPosition === cursorPosition) return
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Is this condition enough? We should also require that the source text is the same, think of copy-paste or changing the source file, etc.

File: src/app.js (hunk 5 above)
Comment (yann300) [reply src/app.js]: Yes now only one sourceunit is managed. but that stuff need to be fixed when we'll add other feature...

File: src/app/editor/contextualListener.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    this._stopWarning()
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Hm, this should perhaps also be called directly whenever we move the cursor out of the current source location.

File: src/app/editor/contextualListener.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      }
+      this.astWalker.walk(compilationResult.sources[source.target].AST, callback)
+    }
+  }
+
+  _warnExpression (node, compilationResult) {
+    var self = this
+    function highlight (id) {
+      if (self._index['ReferencedDeclarations'] && self._index['ReferencedDeclarations'][id]) {
+        var calls = self._index['ReferencedDeclarations'][id]
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: `references` instead of `calls`?

File: src/app/editor/contextualListener.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        for (var call in calls) {
+          self._warn(calls[call].src, compilationResult)
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: We should not only highlight the reference but also the declaration.

Comment (yann300) [reply src/app/editor/contextualListener.js]: the source location of the declaration is a bit tricky to hightlight, cause it's the whole block and not the function name.
though we could highlight the whole block with like 0.2 opacity

File: src/app.js (hunk 1 above)
Comment (chriseth) [reply src/app.js]: I think currently, this is really confusing. It sounds like this would be used for compiler warnings.

Comment (chriseth) [reply src/app.js]: Isn't "highlight" actually more general than "warn"?

File: src/app/editor/contextualListener.js (hunk 4 above)
Comment (chriseth) [reply src/app/editor/contextualListener.js]: Why not `declaration` then?

Comment (chriseth) [reply src/app/editor/contextualListener.js]: The two sides of the "link" should be called `declaration` and `reference`. Neither of them has to be an expression.

File: src/app.js (hunk 5 above)
Comment (chriseth) [reply src/app.js]: Sure, that's fine, I just did not see the check makes sure that `position` and `config.get('currentFile')` reference the same file.

File: src/app/editor/contextualListener.js (hunk 10 above)
Comment (chriseth) [reply src/app/editor/contextualListener.js]: Ah, I see. Perhaps we could also search for a sub-node called "name" inside the node and highlight that?

File: src/app/editor/contextualListener.js (hunk 6 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: no but this is not really linked to the compilation event, the function in setInterval will be executed every 1s and check: if the position of the cursor changed and if there's a compilation result to work with.

File: src/app/editor/contextualListener.js (hunk 7 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: ok so we could just unwarn when content changed

File: src/app/editor/contextualListener.js (hunk 8 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: this is the case

File: src/app/editor/contextualListener.js (hunk 9 above)
Comment (yann300) [reply src/app/editor/contextualListener.js]: yes

File: src/app/editor/contextualListener.js (hunk 6 above)
Comment (chriseth) [reply src/app/editor/contextualListener.js]: ok!

File: src/app.js [hunk 11]
```diff
@@ -383,6 +384,43 @@ function run () {
...
+  // ---------------- ContextualListener -----------------------
+  this._components.contextualListener = new ContextualListener({
+    getCursorPosition: () => {
+      return this._components.editor.getCursorPosition()
+    },
+    getCompilationResult: () => {
+      return compiler.lastCompilationResult
+    },
+    highlight: (position, node) => {
+      if (compiler.lastCompilationResult && compiler.lastCompilationResult.source && compiler.lastCompilationResult.source.target === config.get('currentFile')) {
```
Comment (chriseth) [inline src/app.js]: Please replace `compiler.lastCompilationResult.source.target === config.get('currentFile')` by a check that `position.file` is the file currently shown in the editor.

File: src/app.js [hunk 12]
```diff
@@ -383,6 +384,43 @@ function run () {
...
+        position = offsetToLineColumnConverter.offsetToLineColumn(position, position.file, compiler.lastCompilationResult)
+        var css = 'highlightreference'
+        if (node.children && node.children.length) {
```
Comment (chriseth) [inline src/app.js]: Please add a comment about what this block does.

File: src/app/editor/contextualListener.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+var AstWalker = require('ethereum-remix').util.AstWalker
+
+class ContextualListener {
+  constructor (api, events) {
+    this._api = api
+    this._index = {
+      Declarations: {},
+      FlatReferences: {}
+    }
+    this._events = []
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Please rename to something like `activeHighlights`

File: src/app/editor/contextualListener.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  _highlightItems (cursorPosition, compilationResult) {
+    if (this.currentPosition === cursorPosition) return
+    this._stopHighlighting()
+    this.currentPosition = cursorPosition
+    if (compilationResult && compilationResult.data && compilationResult.source) {
+      var nodes = this.sourceMappingDecoder.nodesAtPosition(null, cursorPosition, compilationResult.data.sources[compilationResult.source.target])
+      if (nodes && nodes.length && nodes[nodes.length - 1]) {
+        this._hightlightExpressions(nodes[nodes.length - 1], compilationResult)
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Typo: high`t`light

File: src/app/editor/contextualListener.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    if (eventId) {
+      this._activeHighlights.push({ eventId, position, fileTarget: compilationResult.source.target })
+    }
+  }
+
+  _hightlightExpressions (node, compilationResult) {
+    var self = this
+    function highlights (id) {
+      if (self._index['Declarations'] && self._index['Declarations'][id]) {
+        var calls = self._index['Declarations'][id]
```
Comment (chriseth) [inline src/app/editor/contextualListener.js]: Please rename `calls`

**Pull request merged**